network/router.go

```diff
@@ -0,0 +1,394 @@
+		if err := c.Close(); err != nil {
+			log.Error(r.address, "having error closing conn to ", remote.Address, ":", err)
+		}
+		// and release one on the waitgroup
+		r.wg.Done()
+	}()
+	address := c.Remote()
+	log.Lvl3(r.address, "Handling new connection to ", remote.Address)
+	for {
+		packet, err := c.Receive()
+		// So the receiver can know about the error
+		packet.SetError(err)
+		packet.From = address
+		packet.ServerIdentity = remote
+
+		// whether the router is closed
+		if r.Closed() {
+			// signal we are done with this go routine.
+			return
+		}
+
+		if err != nil {
+			// something went wrong on this connection
+			log.Lvlf4("%+v got error (%+s) while receiving message", r.id.String(), err)
+
+			if err == ErrClosed || err == ErrEOF || err == ErrTemp {
+				// remote connection closed
+				log.Lvl3(r.address, "handleConn with closed connection: stop (dst=", remote.Address, ")")
+				return
+			}
+			// weird error let's try again
+			log.Lvl3(r.id, "Error with connection", address, "=>", err)
+			continue
+		}
+
+		if err := r.Dispatch(&packet); err != nil {
+			log.Lvl3("Error dispatching:", err)
+		}
+	}
+}
+
+// connection returns the connection associated with this ServerIdentity. Nil if
+// nothing found. It always return the first connection associated.
+func (r *Router) connection(sid ServerIdentityID) Conn {
+	r.connsMut.Lock()
+	defer r.connsMut.Unlock()
+	arr := r.connections[sid]
+	if len(arr) == 0 {
+		return nil
+	}
+	return arr[0]
+}
+
+// registerConnection registers an ServerIdentity for a new connection, mapped with the
+// real physical address of the connection and the connection itself
+// it locks (and unlocks when done):  networkLock
+func (r *Router) registerConnection(remote *ServerIdentity, c Conn) {
+	log.Lvl4(r.address, "Registers", remote.Address)
+	r.connsMut.Lock()
+	defer r.connsMut.Unlock()
+	_, okc := r.connections[remote.ID]
+	if okc {
+		log.Lvl5("Connection already registered. Appending new connection to same identity.")
+	}
+	r.connections[remote.ID] = append(r.connections[remote.ID], c)
+}
+
+func (r *Router) reset() {
+	r.closedMut.Lock()
+	r.isClosed = false
+	r.closedMut.Unlock()
+}
+
+func (r *Router) launchHandleRoutine(dst *ServerIdentity, c Conn) {
+	r.wg.Add(1)
+	go r.handleConn(dst, c)
+}
+
+// Close shuts down all network connections and returns once all processing go
+// routines are done.
+func (r *Router) stopHandling() error {
+	// set the isClosed to true
+	r.close()
+
+	// then close all connections
+	r.connsMut.Lock()
+	for _, arr := range r.connections {
+		// take all connections to close
+		for _, c := range arr {
+			if err := c.Close(); err != nil {
+				log.Error(err)
+			}
+		}
+	}
+	r.connsMut.Unlock()
+
+	// wait for all handleConn to finish
+	r.wg.Wait()
+	return nil
+}
+
+// Closed returns true if the router is closed (or is closing). For a router
+// to be closed means that a call to Stop() must have been made.
+func (r *Router) Closed() bool {
```

Comment:
I'm not sure about this one, I guess it can still be useful to know if the `Router` is closed or not *externally* from the network package ? And also, it makes the code much more clean because I use this `Closed` in a `if` condition, so I must open the lock before the condition and unlock it at two places (inside the `if` and in the `else`) because I don't want this mutex to be locked during the full lifetime of the function.